## wavelets/Snakefile: making files for JG project on admixture and wavelets

## pop_allele_freqs: estimates the minor allele frequency at all SNPs for allopatric maize and mexicana
rule pop_allele_freqs:
    input:
        bams = lambda wildcards: get_all_bams(wildcards.PREFIX), # all input bams
        bais = lambda wildcards: get_all_bais(wildcards.PREFIX), # all input bam indexes
        bam_list = "samples/ALL_byPop/{POP}_bams.list", # bams for focal population
        sites = "local_ancestry/results/thinnedSNPs/{PREFIX}/K{K}/whole_genome.var.sites",
        idx = "local_ancestry/results/thinnedSNPs/{PREFIX}/K{K}/whole_genome.var.sites.idx",
        bin = "local_ancestry/results/thinnedSNPs/{PREFIX}/K{K}/whole_genome.var.sites.bin",
        ref = ref,
        fai = fai
    output:
        mafs = "wavelets/results/alleleFreqs/{PREFIX}/K{K}/{POP}.mafs.gz"
    params:
        p = "med2",
        out_prefix = lambda wildcards: "wavelets/results/alleleFreqs/" + wildcards.PREFIX + "/K" + wildcards.K + "/" + wildcards.POP
    shadow:
        "minimal"
    conda:
        "../envs/environment.yaml"
    threads:
        1
    resources:
        time_min = lambda wildcards, attempt: attempt * 24 * 60,
        mem = lambda wildcards, attempt: attempt * 16
    shell:
        "angsd -out {params.out_prefix} "
        "-r {params.region} "
        "-sites {input.sites} "
        "-ref {input.ref} "
        "-bam {input.bam_list} "
        "-remove_bads 1 "
        "-minMapQ 30 -minQ 20 "
        "-doMajorMinor 3 "
        "-P 1 "
        "-baq 2 "
        "-GL 1 -doMaf 1"
