# pipeline to call SNPs
workdir: path_hilo + "variant_sites/" # some rules need an absolute path to create a 'shadow' copy of files outside of current working directory

name_all = "HILO_MAIZE55"
#name_all = "TEST"

# rule to get random regions from the genome
# rule to get bams list of included samples

# calculate depth of sequencing coverage for the combined sample, only including bases meeting minQ >20
rule calc_depth:
    input:
        bams = path_hilo + "samples/" + name_all + "_bams.list",
        regions = path_hilo + "data/refMaize/random_regions/N1000.L100.regions"
    output:
        "results/depthCov/N1000.L100.regions/" + name_all + ".Q20.depthGlobal"
    params:
        p = "bigmemm",
        out_prefix = "results/depthCov/N1000.L100.regions/" + name_all + ".Q20"
#    conda:
#        "../envs/environment.yaml"
    shadow:
        "minimal"
    threads:
        4
    resources:
        time_min = lambda wildcards, attempt: attempt * 12 * 60,
        mem = 32
    shell:
        "angsd -out {params.out_prefix} "
        "-rf {input.regions} "
        "-bam {input.bams} "
        "-remove_bads 1 -minMapQ 30 -minQ 20 "
        "-doCounts 1 -doDepth 1 -maxDepth 10000"

# calculate total depth from the .Q20.depthGlobal file to use as input for 2x depth 
