## variant_sites/Snakefile: pipeline to call SNPs

workdir: path_hilo
# note: working directory is hilo/ and all file inputs/outputs below are relative to that working directory

## thin_GL_4PCA: thins sites in beagle genotype likelihood files for all regions to >0.01cM spacing (~10kb)
## and returns one concatenated GL file for input to PCAngsd and NGSadmix
rule thin_GL_4PCA:
    input: # input all GL files, corresponding rpos files, and the master list of all regions
        gl = expand("variant_sites/results/" + prefix_all + "/{REGION}.beagle.gz",
        #REGION = ["region_0", "region_1", "region_300"]),
        REGION=list(regions_dict.keys())),
        rpos = expand("variant_sites/results/" + prefix_all + "/{REGION}.rpos",
        #REGION = ["region_0", "region_1", "region_300"]),
        REGION=list(regions_dict.keys())),
        regions = "data/refMaize/divide_5Mb/ALL_regions.list"
        #regions = "test/TEST2_regions.list"
    output: # thinned GL file for whole genome
        gl = "global_ancestry/results/thinnedSNPs/" + prefix_all + "/whole_genome.beagle.gz"
    params:
        p = "med2",
        prefix_all = prefix_all,
        min_cM = 0.01 # minimum spacing
    shadow:
        "minimal"
    resources:
        time_min = lambda wildcards, attempt: attempt * 6 * 60,
        mem = lambda wildcards, attempt: attempt * 4
    conda:
        "../envs/environment.yaml"
    script:
        "thin_GL_4PCA.R"

